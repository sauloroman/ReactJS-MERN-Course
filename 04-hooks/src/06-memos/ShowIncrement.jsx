import React from "react"

export const ShowIncrement = React.memo(({ incrementFn }) => {
  console.log('Render ShowIncrement')
  return (
    <button className="btn" onClick={ () => incrementFn(2) }>Incrementar contador</button>
  )
})
