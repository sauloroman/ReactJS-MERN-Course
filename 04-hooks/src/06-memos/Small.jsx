import React from 'react'

export const Small = React.memo(({ value }) => {
  console.log('Render Small')
  return (
    <div className="counter">
      <p className="counter__value">{value}</p>
      <div className="counter__name">Contador</div>
    </div>
  )
})
