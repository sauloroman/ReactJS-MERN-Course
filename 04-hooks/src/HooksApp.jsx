// import { CounterApp } from "./01-useState/CounterApp"
// import { CounterAppWithCustomHook } from "./01-useState/CounterAppWithCustomHook"
// import { SimpleForm } from "./02-useEffect/SimpleForm"
// import { FormWithCustomHook } from "./02-useEffect/FormWithCustomHook"
// import { MultipleCustomHooks } from "./03-examples/MultipleCustomHooks"
// import { FocusScreen } from "./04-useRef/FocusScreen"
// import { Layout } from "./05-useLayoutEffect/Layout"
// import { MemoHook } from "./06-memos/MemoHook"
// import { Memorize } from "./06-memos/Memorize"
// import { CallbackHook } from "./06-memos/CallbackHook"
// import { Padre } from "./07-tarea-memo/Padre"
import { MultipleCustomHook } from "./03-examples-new/MultipleCustomHook"

export const HooksApp = () => {
  return (
    <div className="container">
      <MultipleCustomHook />
    </div>
  )
}
