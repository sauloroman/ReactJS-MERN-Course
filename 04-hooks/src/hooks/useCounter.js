import { useState } from "react"

export const useCounter = ( initialValue = 1 ) => {

  const [counter, setCounter] = useState(initialValue)

  const onIncrementCounter = ( value = 1 ) => {
    setCounter( counter => counter + value )
  }

  const onDecrementCounter = ( value = 1 ) => {
    if ( counter === 0 ) return
    setCounter( counter => counter - value )
  }

  const onResetCounter = () => setCounter(initialValue)

  return {
    // State
    counter,

    // Methods
    onIncrementCounter,
    onDecrementCounter,
    onResetCounter
  }

}