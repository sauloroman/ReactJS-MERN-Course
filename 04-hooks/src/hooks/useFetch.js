import { useEffect } from "react"
import { useState } from "react"

export const useFetch = ( url = '' ) => {

  const [state, setState] = useState({
    data: null,
    hasError: null,
    isLoading: false,
  })

  useEffect(() => {
    getData()
  }, [ url ])

  const getData = async () => {

    setState({
      ...state,
      isLoading: true,
    })

    try {
      
      const res = await fetch( url );
      const data = await res.json()
    
      setState({
        ...state,
        data: data,
        isLoading: false,
      })

    } catch (error) {
      setState({
        ...state,
        hasError: error,
        isLoading: false,
      })
    }

  }

  return {
    data: state.data,
    hasError: state.hasError,
    isLoading: state.isLoading,
  }

}