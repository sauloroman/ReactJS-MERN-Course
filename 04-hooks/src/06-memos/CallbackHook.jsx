import { useCallback, useState } from "react"
import { ShowIncrement } from "./ShowIncrement"

export const CallbackHook = () => {

  const [counter, setCounter] = useState(10)

  const increment = useCallback(( value ) => setCounter( counter => counter + value ), []);

  // const increment = ( value ) => {
  //   setCounter( counter => counter + value )
  // }

  return (
    <>
      <h1 className="heading">UseCallback</h1>

      <p className="counter">Counter: {counter}</p>

      <ShowIncrement incrementFn={ increment } />
    </>
  )
}
