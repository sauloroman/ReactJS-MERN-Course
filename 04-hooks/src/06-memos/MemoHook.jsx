import { useMemo, useState } from "react"
import { useCounter } from "../hooks"

const heavyStuff = ( iterationTimes = 1000 ) => {

  for( let i = 1; i <= iterationTimes; i++ ) {
    console.log('Ahi vamos')
  }

  return `${iterationTimes} iteraciones hechas`
}

export const MemoHook = () => {
  
  const [isShown, setIsShown] = useState(false)
  const { counter, onIncrementCounter } = useCounter(4000)
  const memorizedValue = useMemo(() => heavyStuff(counter), [counter])

  return (
    <>
      <h2 className="heading">UseMemo</h2>

      <p className="counter__value">Counter: {counter}</p>

      <button className="btn" onClick={ () => onIncrementCounter() }>+1</button>

      <button
        className="btn"
        onClick={ () => setIsShown( !isShown )}
      >Show | Hide {JSON.stringify(isShown)}</button>

      <span>{memorizedValue}</span>
    </>
  )
}
