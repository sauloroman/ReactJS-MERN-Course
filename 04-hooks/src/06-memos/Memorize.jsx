import { useState } from 'react'
import { useCounter } from '../hooks'
import { Small } from './Small'

export const Memorize = () => {
  
  const { counter, onIncrementCounter } = useCounter(10)
  const [isShown, setIsShown] = useState(false)

  return (
    <>
      <h1 className="heading">Memo Function</h1>

      <Small value={counter} />

      <button className="btn" onClick={() => onIncrementCounter()}>+1</button>

      <button onClick={() => setIsShown(!isShown)} className='btn'>Shown | Hidden {JSON.stringify( isShown )}</button>
    </>
  )
}
