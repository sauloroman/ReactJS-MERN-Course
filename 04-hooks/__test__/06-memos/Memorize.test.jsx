import { fireEvent, render } from '@testing-library/react'
import { useCounter } from '../../src/hooks'
import { Memorize } from '../../src/06-memos/Memorize'

jest.mock('../../src/hooks/useCounter.js')

describe('Tests in Memorize Component', () => {

  const onIncrementCounterMock = jest.fn()

  useCounter.mockReturnValue({
    counter: 1,
    onIncrementCounter: onIncrementCounterMock
  })

  test('it must match with snapshot', () => {
    const { container } = render(<Memorize />)
    expect( container ).toMatchSnapshot()
  })

  test('it must call onIncrementCounter function', () => {
    const { getByText } = render(<Memorize />)

    const buttonToAdd = getByText('+1')
    fireEvent.click( buttonToAdd )

    expect( onIncrementCounterMock ).toHaveBeenCalled()
  })

  test('it must change the value of the state and the button text', () => {

    const { getByTestId } = render(<Memorize />)

    const buttonStatus = getByTestId('button-change-status')
    
    fireEvent.click( buttonStatus )
    expect( buttonStatus.textContent ).toContain('Shown | Hidden true')
    
    fireEvent.click( buttonStatus )
    expect( buttonStatus.textContent ).toContain('Shown | Hidden false')

  })

})  