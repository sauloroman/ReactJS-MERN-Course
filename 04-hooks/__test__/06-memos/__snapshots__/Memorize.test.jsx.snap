// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in Memorize Component it must match with snapshot 1`] = `
<div>
  <h1
    class="heading"
  >
    Memo Function
  </h1>
  <div
    class="counter"
  >
    <p
      class="counter__value"
    >
      1
    </p>
    <div
      class="counter__name"
    >
      Contador
    </div>
  </div>
  <button
    class="btn"
  >
    +1
  </button>
  <button
    class="btn"
    data-testid="button-change-status"
  >
    Shown | Hidden 
    false
  </button>
</div>
`;
